Flush batch cause panic

### Issue description
Using Flush() method in clickhouse.Batch causes panic

### Error log
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x895d0b]

goroutine 12 [running]:
github.com/ClickHouse/ch-go/proto.(*Buffer).PutUInt8(...)
	/go/pkg/mod/github.com/!click!house/ch-go@v0.47.3/proto/buffer.go:97
github.com/ClickHouse/ch-go/proto.(*Buffer).PutByte(...)
	/go/pkg/mod/github.com/!click!house/ch-go@v0.47.3/proto/buffer.go:82
github.com/ClickHouse/clickhouse-go/v2.(*connect).sendData(0xc000146270, 0xc00011e1c0?, {0x0, 0x0})
	/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.3.0/conn.go:171 +0xcb
github.com/ClickHouse/clickhouse-go/v2.(*batch).Flush(0xc00014e280)
	/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.3.0/conn_batch.go:165 +0x79
```

### Configuration
*OS:* `Alpine Linux 3.16`

*Interface:* `native`

*Go version:* 1.19

*ClickHouse Server version:* 22.9.3.18

Still panicking with Send() method. 

Sample code 
```
if err := c.batch.Append(/* some data */); err != nil {
	return fmt.Errorf("append to batch: %w", err)
}

c.leftRows--
if c.leftRows == 0 {
	if err = c.batch.Send(); err != nil {
		return fmt.Errorf("send batch: %w", err)
	}

	c.batch, err = c.conn.PrepareBatch(context.Background(), batchQuery)
	if err != nil {
		return fmt.Errorf("reinit batch: %w", err)
	}

	c.logger.Debug("send batch")
	c.leftRows = c.BatchCount
}

return nil
```

Trace log 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x895d0b]

goroutine 66 [running]:
github.com/ClickHouse/ch-go/proto.(*Buffer).PutUInt8(...)
	/go/pkg/mod/github.com/!click!house/ch-go@v0.47.3/proto/buffer.go:97
github.com/ClickHouse/ch-go/proto.(*Buffer).PutByte(...)
	/go/pkg/mod/github.com/!click!house/ch-go@v0.47.3/proto/buffer.go:82
github.com/ClickHouse/clickhouse-go/v2.(*connect).sendData(0xc000406a90, 0xc0?, {0x0, 0x0})
	/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.3.0/conn.go:171 +0xcb
github.com/ClickHouse/clickhouse-go/v2.(*batch).Send(0xc0003b0000)
	/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.3.0/conn_batch.go:144 +0xe5
```

I'm getting the same during `.Send()`
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x14a260b]

goroutine 1 [running]:
github.com/ClickHouse/ch-go/proto.(*Buffer).PutUInt8(...)
	/Users/vnazarenko/code/go/pkg/mod/github.com/!click!house/ch-go@v0.47.3/proto/buffer.go:97
github.com/ClickHouse/ch-go/proto.(*Buffer).PutByte(...)
	/Users/vnazarenko/code/go/pkg/mod/github.com/!click!house/ch-go@v0.47.3/proto/buffer.go:82
github.com/ClickHouse/clickhouse-go/v2.(*connect).sendData(0xc0004f15f0, 0x17c27e0?, {0x0, 0x0})
	/Users/vnazarenko/code/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.3.0/conn.go:171 +0xcb
github.com/ClickHouse/clickhouse-go/v2.(*batch).Send(0xc0006a6550)
	/Users/vnazarenko/code/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.3.0/conn_batch.go:148 +0x12d
main.(*CassandraUsersIterator).IterateUsersInBatchesCh(0xc00030fd50, 0xc0001a2008?, 0xc00063fdf8?, 0x8000000000000000)
	/Users/vnazarenko/code/go/src/gitlab.com/ipl/cacheinfo/cmd/cache_users_iterator/clickhouse_saver.go:306 +0x72d
main.main()
	/Users/vnazarenko/code/go/src/gitlab.com/ipl/cacheinfo/cmd/cache_users_iterator/main.go:188 +0x986
```

> @gingerwizard, is clickhouse connection a concurrent safe object?
> 
> we are using ch-go in the following pattern, and is frequently seeing segmentation violation error with latest release.
> 
> ```
> func writeRows(prepareSQL string, rows model.Rows, conn clickhouse.Conn) (err error) {
> 	var batch driver.Batch
> 	if batch, err = conn.PrepareBatch(context.Background(), prepareSQL)
> 	defer func() {
> 		_ = batch.Abort()
> 	}()
> 	for _, row := range rows {
> 		batch.Append((*row)...)
> 	}
> 	return err = batch.Send()
> }
> 
> func Run() {
> 	...
> 	conn = clickhouse.Open(&sc.opts)
> 	todo, done := int64(len(items)), int64(-1)
> 	var wg sync.WaitGroup
> 	wg.Add(10)
> 	for i := 0; i < 10; i++ {
> 		go func() {
> 			for {
> 				index := atomic.AddInt64(&done, 1)
> 				if index >= todo {
> 					wg.Done()
> 					break
> 				}
> 				writeRows(items[index].sql, items[index].rows, conn)
> 				items[index] = nil
> 			}
> 		}()
> 	}
> 	wg.Wait()
> 	...
> }
> ```

Attaching some of the panic log for your reference:
```
goroutine 242 [running]:
github.com/ClickHouse/ch-go/proto.(*Buffer).PutUVarInt(...)
        /root/go/pkg/mod/github.com/!click!house/ch-go@v0.50.0/proto/buffer.go:72
github.com/ClickHouse/ch-go/proto.(*Buffer).PutLen(...)
        /root/go/pkg/mod/github.com/!click!house/ch-go@v0.50.0/proto/buffer.go:87
github.com/ClickHouse/ch-go/proto.(*Buffer).PutString(...)
        /root/go/pkg/mod/github.com/!click!house/ch-go@v0.50.0/proto/buffer.go:92
github.com/ClickHouse/clickhouse-go/v2/lib/proto.(*Query).Encode(0xc8b1f876c0?, 0x0?, 0xd4b5?)
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/lib/proto/query.go:45 +0xb6
github.com/ClickHouse/clickhouse-go/v2.(*connect).sendQuery(0xc8b5dfd6c0, {0xccc644c750, 0xca}, 0xc6d4cfe420)
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/conn_send_query.go:38 +0x5eb
github.com/ClickHouse/clickhouse-go/v2.(*connect).prepareBatch(0xc8b5dfd6c0, {0x14fd3a0?, 0xc0000460d8}, {0xc000033c70?, 0x0?}, 0xc9a8448d20)
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/conn_batch.go:55 +0x325
github.com/ClickHouse/clickhouse-go/v2.(*clickhouse).PrepareBatch(0xc5c29fe8c0, {0x14fd3a0, 0xc0000460d8}, {0xc000033c70, 0xc3})
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/clickhouse.go:152 +0x99
github.com/housepower/clickhouse_sinker/output.writeRows({0xc000033c70, 0xc3}, {0xcad8100000, 0x10001, 0x6e65757165735f39?}, 0x0, 0xf, {0x1509830, 0xc5c29fe8c0})
        /usr/local/src/sinker/clickhouse_sinker/output/clickhouse_util.go:36 +0xb6
```
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xf202c4]

goroutine 187 [running]:
github.com/ClickHouse/ch-go/proto.(*Buffer).PutUInt8(...)
        /root/go/pkg/mod/github.com/!click!house/ch-go@v0.50.0/proto/buffer.go:97
github.com/ClickHouse/ch-go/proto.(*Buffer).PutByte(...)
        /root/go/pkg/mod/github.com/!click!house/ch-go@v0.50.0/proto/buffer.go:82
github.com/ClickHouse/clickhouse-go/v2.(*connect).sendQuery(0xc536af1a00, {0xc49c2be8f0, 0xca}, 0xc96edaa160)
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/conn_send_query.go:28 +0x104
github.com/ClickHouse/clickhouse-go/v2.(*connect).prepareBatch(0xc536af1a00, {0x14fd3a0?, 0xc000198000}, {0xc0002e2d00?, 0x0?}, 0xc96fffebf0)
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/conn_batch.go:55 +0x325
github.com/ClickHouse/clickhouse-go/v2.(*clickhouse).PrepareBatch(0xc4445f6f80, {0x14fd3a0, 0xc000198000}, {0xc0002e2d00, 0xc3})
        /root/go/pkg/mod/github.com/!click!house/clickhouse-go/v2@v2.4.2/clickhouse.go:152 +0x99
github.com/housepower/clickhouse_sinker/output.writeRows({0xc0002e2d00, 0xc3}, {0xcab78cc000, 0x10001, 0x363?}, 0x0, 0xf, {0x1509830, 0xc4445f6f80})
        /usr/local/src/sinker/clickhouse_sinker/output/clickhouse_util.go:36 +0xb6
```

--- a/conn_batch.go
+++ b/conn_batch.go
@@ -35,6 +35,11 @@ var splitInsertRe = regexp.MustCompile(`(?i)\sVALUES\s*\(`)
 var columnMatch = regexp.MustCompile(`.*\((?P<Columns>.+)\)$`)
 
 func (c *connect) prepareBatch(ctx context.Context, query string, release func(*connect, error)) (driver.Batch, error) {
+	//defer func() {
+	//	if err := recover(); err != nil {
+	//		fmt.Printf("panic occurred on %d:\n", c.num)
+	//	}
+	//}()
 	query = splitInsertRe.Split(query, -1)[0]
 	colMatch := columnMatch.FindStringSubmatch(query)
 	var columns []string
@@ -69,24 +74,31 @@ func (c *connect) prepareBatch(ctx context.Context, query string, release func(*
 		return nil, err
 	}
 	return &batch{
-		ctx:   ctx,
-		conn:  c,
-		block: block,
-		release: func(err error) {
-			release(c, err)
-		},
-		onProcess: onProcess,
+		ctx:         ctx,
+		conn:        c,
+		block:       block,
+		released:    false,
+		connRelease: release,
+		onProcess:   onProcess,
 	}, nil
 }
 
 type batch struct {
-	err       error
-	ctx       context.Context
-	conn      *connect
-	sent      bool
-	block     *proto.Block
-	release   func(error)
-	onProcess *onProcess
+	err         error
+	ctx         context.Context
+	conn        *connect
+	sent        bool
+	released    bool
+	block       *proto.Block
+	connRelease func(*connect, error)
+	onProcess   *onProcess
+}
+
+func (b *batch) release(err error) {
+	if !b.released {
+		b.released = true
+		b.connRelease(b.conn, err)
+	}
 }
 
 func (b *batch) Abort() error {
@@ -104,7 +116,9 @@ func (b *batch) Append(v ...interface{}) error {
 	if b.sent {
 		return ErrBatchAlreadySent
 	}
-	//
+	if b.err != nil {
+		return b.err
+	}
 	if err := b.block.Append(v...); err != nil {
 		b.err = errors.Wrap(ErrBatchInvalid, err.Error())
 		b.release(err)
@@ -114,6 +128,9 @@ func (b *batch) Append(v ...interface{}) error {
 }
 
 func (b *batch) AppendStruct(v interface{}) error {
+	if b.err != nil {
+		return b.err
+	}
 	values, err := b.conn.structMap.Map("AppendStruct", b.block.ColumnsNames(), v, false)
 	if err != nil {
 		return err

--- a/tests/issues/798_test.go
+++ b/tests/issues/798_test.go
@@ -5,6 +5,8 @@ import (
 	"github.com/ClickHouse/clickhouse-go/v2"
 	clickhouse_tests "github.com/ClickHouse/clickhouse-go/v2/tests"
 	"github.com/stretchr/testify/require"
+	"sync"
+	"sync/atomic"
 	"testing"
 )
 
@@ -50,4 +52,61 @@ func Test798(t *testing.T) {
 	require.NoError(t, batch.Append(true, false, []bool{true, false, true}))
 	require.NoError(t, batch.Send())
 	require.ErrorIs(t, batch.Append(true, false, []bool{true, false, true}), clickhouse.ErrBatchAlreadySent)
+}
+
+func writeRows(prepareSQL string, rows [][]interface{}, conn clickhouse.Conn) (err error) {
+	batch, err := conn.PrepareBatch(context.Background(), prepareSQL)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if batch != nil {
+			_ = batch.Abort()
+		}
+	}()
+	for _, row := range rows {
+		batch.Append(row...)
+	}
+	return batch.Send()
+}
+
+func Test798Concurrent(t *testing.T) {
+
+	var (
+		conn, err = clickhouse_tests.GetConnection("issues", clickhouse.Settings{
+			"max_execution_time": 60,
+		}, nil, &clickhouse.Compression{
+			Method: clickhouse.CompressionLZ4,
+		})
+	)
+	ctx := context.Background()
+	const ddl = `
+			CREATE TABLE test_issue_798 (
+				  Col1 Bool
+				, Col2 Bool
+			) Engine MergeTree() ORDER BY tuple()
+		`
+	defer func() {
+		conn.Exec(ctx, "DROP TABLE IF EXISTS test_issue_798")
+	}()
+	require.NoError(t, conn.Exec(ctx, ddl))
+
+	require.NoError(t, err)
+	todo, done := int64(1000), int64(-1)
+	var wg sync.WaitGroup
+	wg.Add(10)
+	for i := 0; i < 10; i++ {
+		go func() {
+			for {
+				index := atomic.AddInt64(&done, 1)
+				if index >= todo {
+					wg.Done()
+					break
+				}
+				writeRows("INSERT INTO test_issue_798", [][]interface{}{{true, false}, {false, true}}, conn)
+			}
+		}()
+	}
+	wg.Wait()
+
 }

